Neos.ContentRepository/Classes/Domain/Projection/Content/NodeInterface.php

```diff
@@ -0,0 +1,124 @@
+<?php
+namespace Neos\ContentRepository\Domain\Projection\Content;
+
+/*
+ * This file is part of the Neos.ContentRepository package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\ContentRepository\Domain\Model\NodeType;
+use Neos\ContentRepository\Domain\ValueObject\NodeIdentifier;
+use Neos\ContentRepository\Domain\ValueObject\ContentStreamIdentifier;
+use Neos\ContentRepository\DimensionSpace\DimensionSpace\DimensionSpacePoint;
+use Neos\ContentRepository\Domain\ValueObject\NodeAggregateIdentifier;
+use Neos\ContentRepository\Domain\ValueObject\NodeName;
+use Neos\ContentRepository\Domain\ValueObject\NodeTypeName;
+use Neos\Cache\CacheAwareInterface;
+use Neos\ContentRepository\Domain\ValueObject\PropertyCollectionInterface;
+
+/**
+ * The "new" Event-Sourced NodeInterface. Supersedes the old Neos\ContentRepository\Domain\Model\NodeInterface.
+ *
+ * !! Reference resolving NOT in NodeInterface
+ *
+ * Immutable. Read-only. Detached from storage.
+ */
+interface NodeInterface extends CacheAwareInterface
+{
+    /**
+     * @return ContentStreamIdentifier
+     */
+    public function getContentStreamIdentifier(): ContentStreamIdentifier;
+
+    /**
+     * returns the DimensionSpacePoint the node was *requested in*, i.e. one of the DimensionSpacePoints
+     * this node is visible in. If you need the DimensionSpacePoint where the node is actually at home,
+     * see getOriginDimensionSpacePoint()
+     *
+     * @return DimensionSpacePoint
+     */
+    public function getDimensionSpacePoint(): DimensionSpacePoint;
+
+    /**
+     * @return NodeAggregateIdentifier
+     */
+    public function getNodeAggregateIdentifier(): NodeAggregateIdentifier;
+
+    /**
+     * @return NodeIdentifier
+     */
+    public function getNodeIdentifier(): NodeIdentifier;
+
+    /**
+     * @return NodeTypeName
+     */
+    public function getNodeTypeName(): NodeTypeName;
+
+    /**
+     * @return NodeType|null
+     */
+    public function getNodeType(): ?NodeType;
+
+    /**
+     * @return NodeName
+     */
+    public function getNodeName(): NodeName;
+
+    /**
+     * returns the DimensionSpacePoint the node is at home in. Usually needed to address a Node in a NodeAggregate
+     * in order to update it.
+     *
+     * @return DimensionSpacePoint
+     */
+    public function getOriginDimensionSpacePoint(): DimensionSpacePoint;
+
+    /**
+     * Returns all properties of this node. References are NOT part of this API; there you need to check getReference() and getReferences()
+     *
+     * @return PropertyCollectionInterface Property values, indexed by their name
+     * @api
+     */
+    public function getProperties(): PropertyCollectionInterface;
```

Comment:
I'd say yes. In different implementations the interface can be implemented differently.

Reply:
I'd be interested in the use cases. I would think that the node properties are just a "PHP array on steroids". Additional logic (like special conversions) are IMO better located in the custom Node implementation or (depending on the use case) in the consuming part

Reply:
@bwaidelich in Neos, we need the "PHP serialized object to entity conversion"; and I'd like to do this lazily, e.g. when reading an `image` property. That's why I wanted to encapsulate this in an implementation of `PropertyCollectionInterface`; instead of eagerly doing it on *any* property access.

Reply:
This can be helpful for imported, schemaless nodes where no corresponding property declaration via NodeTypes.yaml is available